sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java

```diff
@@ -788,6 +790,24 @@ static Pair<DataSource, Filtration> getFiltration(
   {
     if (!canUseIntervalFiltering(dataSource)) {
       return Pair.of(dataSource, toFiltration(filter, virtualColumnRegistry.getFullRowSignature(), false));
+    } else if (dataSource instanceof UnnestDataSource) {
+      // UnnestDataSource can have another unnest data source
+      // join datasource, filtered data source, etc as base
+      Pair<DataSource, Filtration> pair = getFiltration(
+          ((UnnestDataSource) dataSource).getBase(),
```

Comment:
Can you please add test case for nested UnnestDatasources ? just want to make sure that correct interval populates correctly for 2-3 nested case.

Reply:
Added more tests with 2 and 3 levels of nesting

Reply:
We are not checking if the underlying data source can use interval filtering. What's the behavior when we replace the table data source with something like a lookup and do an unnest on that? We shouldn't call `optimize` in that case since lookups can't use interval filtering.

Why is the `UNNEST` datasource not triggering the final `else` branch in this scenario if the underlying table `canUseIntervalFiltering` and the very first if branch if the data source can't?

Reply:
We have pushed the filter into the filteredDataSource, so if the last else is triggered, the filter is null and the filtration is not applied and query segment spec of eternity takes over. Regarding the comment on datasource using interval filtering, I'll add a case under unnest to check if the base data source can use interval filtering and trigger the same logic as the last else

Reply:
in the original code before the unnest handling this would appear as
```
datasource = UnnestDS(FilteredDS(foo,timeFilter))
filter = null
```

The presence of a filter would always trigger a filter data source. So that flow was not able to handle this without the special handling for UnnestDataSource which just delegates to its children and the Filter data source which uses the filter present with it